src/microceph_remote.py

```diff
@@ -0,0 +1,293 @@
+        """Configure event handlers for an ceph-nfs-client interface."""
+        logger.debug(f"Setting up {self.relation_name} event handler")
+
+        microceph_remote = MicroCephRemote(self.charm, self.relation_name)
+
+        self.framework.observe(microceph_remote.on.microceph_remote_departed, self._on_departed)
+        self.framework.observe(microceph_remote.on.microceph_remote_reconcile, self._on_reconcile)
+        self.framework.observe(
+            microceph_remote.on.microceph_remote_update_remote, self._on_update_remote
+        )
+
+        return microceph_remote
+
+    @property
+    def ready(self) -> bool:
+        """Property: ready check."""
+        logger.debug(f"Reporting {self.relation_name} ready")
+        return True
+
+    def _on_departed(self, event):
+        """Handle integration cleanup."""
+        logger.debug("Handling remote departed event")
+        remote_relation_data = event.relation.data.get(event.relation.app)
+        remote_site_name = remote_relation_data.get(RemoteRelationDataKeys.site_name.value, None)
+
+        remove_remote_cluster(remote_site_name)
+
+    def _on_reconcile(self, event):
+        logger.debug("Handling remote reconcile event")
+        # fetch remote app data
+        remote_relation_data = event.relation.data.get(event.relation.app, None)
+        remote_site_name = remote_relation_data.get(RemoteRelationDataKeys.site_name.value, None)
+        remote_token = remote_relation_data.get(RemoteRelationDataKeys.token.value, None)
+
+        if remote_site_name and remote_token:
+            import_remote_cluster(
+                local_name=self.charm.model.config.get("site-name"),
+                remote_name=remote_site_name,
+                remote_token=remote_token,
+            )
+
+    def _on_update_remote(self, event):
+        logger.debug("Handling remote update event")
+
+        local_relation_data = event.relation.data.get(self.charm.app)
+        remote_relation_data = event.relation.data.get(event.relation.app)
+
+        local_site_name = local_relation_data.get(RemoteRelationDataKeys.site_name.value, None)
+        if not local_site_name:
+            local_site_name = self.charm.model.config.get("site-name")
+            logger.debug("Updating local site name(%s) in remote relation data", local_site_name)
+            local_relation_data.update({RemoteRelationDataKeys.site_name.value: local_site_name})
+
+        local_token = local_relation_data.get(RemoteRelationDataKeys.token.value, None)
+        remote_site_name = remote_relation_data.get(RemoteRelationDataKeys.site_name.value, None)
+
+        logger.debug(
+            "For Remote site: %s, isLocalToken=%s",
+            remote_site_name,
+            local_token is not None,
+        )
+        # remote site name is required to generate token
+        if not local_token and remote_site_name:
+            new_token = get_cluster_export_token(remote_site_name)
```

Comment:
Ok so I have to admit that the event handling machinery is confusing to me,  please bear with me :-) 
But afaict _on_update_remote() is triggered if the remote peer relation changed right? Does that mean we re-generate a token if the remote cluster is up- or downscaled?

Reply:
```
    microceph_remote_reconcile = EventSource(MicrocephRemoteReconcileEvent)
    microceph_remote_update_remote = EventSource(MicrocephRemoteUpdateRemoteEvent)
```
Here, To absorb the updates in remote we emit the reconcile event, and to propagate updates to remote cluster we trigger update_remote. All of it is done by updating the relation databag.

The `cluster export` and `remote import` CLI commands in microceph are idempotent. So if new monitors are added in the cluster, we do not need to check anything we can basically just re-export the cluster token and re-import it on remote. Juju by itself only triggers remote event if the relation databag has changed.